python/ray/data/_internal/datasource/iceberg_datasink.py

```diff
@@ -0,0 +1,161 @@
+        else:
+            self._catalog_name = "default"
+
+        self._uuid: str = None
+        self._io: FileIO = None
+        self._txn: Transaction = None
+        self._table_metadata: TableMetadata = None
+
+    # Since iceberg transaction is not pickle-able, because of the table and catalog properties
+    # we need to exclude the transaction object during serialization and deserialization during pickle
+    def __getstate__(self) -> dict:
+        """Exclude `_txn` during pickling."""
+        state = self.__dict__.copy()
+        del state["_txn"]
+        return state
+
+    def __setstate__(self, state: dict) -> None:
+        self.__dict__.update(state)
+        self._txn = None
+
+    def _get_catalog(self) -> "Catalog":
+        from pyiceberg import catalog
+
+        return catalog.load_catalog(self._catalog_name, **self._catalog_kwargs)
+
+    def on_write_start(self) -> None:
+        """Prepare for the transaction"""
+        from pyiceberg.table import PropertyUtil, TableProperties
+
+        catalog = self._get_catalog()
+        table = catalog.load_table(self.table_identifier)
+        self._txn = table.transaction()
+        self._io = self._txn._table.io
+        self._table_metadata = self._txn.table_metadata
+        self._uuid = uuid.uuid4()
+
+        if unsupported_partitions := [
+            field
+            for field in self._table_metadata.spec().fields
+            if not field.transform.supports_pyarrow_transform
+        ]:
+            raise ValueError(
+                f"Not all partition types are supported for writes. Following partitions cannot be written using pyarrow: {unsupported_partitions}."
+            )
+
+        self._manifest_merge_enabled = PropertyUtil.property_as_bool(
+            self._table_metadata.properties,
+            TableProperties.MANIFEST_MERGE_ENABLED,
+            TableProperties.MANIFEST_MERGE_ENABLED_DEFAULT,
+        )
+
+    def write(
+        self, blocks: Iterable[Block], ctx: TaskContext
+    ) -> WriteResult[List["DataFile"]]:
+        from pyiceberg.io.pyarrow import (
+            _check_pyarrow_schema_compatible,
+            _dataframe_to_data_files,
+        )
+        from pyiceberg.table import DOWNCAST_NS_TIMESTAMP_TO_US_ON_WRITE
+        from pyiceberg.utils.config import Config
+
+        data_files_list: WriteResult[List["DataFile"]] = []
+        for block in blocks:
+            pa_table = BlockAccessor.for_block(block).to_arrow()
+
+            downcast_ns_timestamp_to_us = (
+                Config().get_bool(DOWNCAST_NS_TIMESTAMP_TO_US_ON_WRITE) or False
+            )
+            _check_pyarrow_schema_compatible(
+                self._table_metadata.schema(),
+                provided_schema=pa_table.schema,
+                downcast_ns_timestamp_to_us=downcast_ns_timestamp_to_us,
+            )
+
+            if pa_table.shape[0] <= 0:
+                continue
+
+            data_files = _dataframe_to_data_files(
```

Comment:
does the actually writing happen in `_dataframe_to_data_files`, and the `data_files` only contains metadata?